src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsDevice.cpp

```diff
@@ -0,0 +1,524 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#include <NovelRT/Experimental/Graphics/Vulkan/Graphics.Vulkan.h>
+
+namespace NovelRT::Experimental::Graphics::Vulkan
+{
+    VulkanGraphicsDevice::VulkanGraphicsDevice(const std::shared_ptr<VulkanGraphicsAdapter>& adapter,
+                                               const std::shared_ptr<VulkanGraphicsSurfaceContext>& surfaceContext,
+                                               int32_t contextCount)
+        : GraphicsDevice(adapter, std::static_pointer_cast<GraphicsSurfaceContext>(surfaceContext)),
+          _presentCompletionFence([&]() {
+              std::shared_ptr<VulkanGraphicsFence> ptr = std::make_shared<VulkanGraphicsFence>(
+                  std::dynamic_pointer_cast<VulkanGraphicsDevice>(shared_from_this()));
+              ptr->Reset();
+              return ptr;
+          }),
+          _contexts([&, contextCount]() { return CreateGraphicsContexts(contextCount); }),
+          _contextPtrs([&]() {
+              std::vector<std::shared_ptr<GraphicsContext>> ptrs{};
+              ptrs.reserve(_contexts.getActual().size());
+
+              for (auto&& context : _contexts.getActual())
+              {
+                  ptrs.emplace_back(std::dynamic_pointer_cast<VulkanGraphicsContext>(context->shared_from_this()));
+              }
+
+              return ptrs;
+          }),
+          _logger(LoggingService(NovelRT::Utilities::Misc::CONSOLE_LOG_GFX)),
+          _surface(GetSurfaceContext()->GetVulkanSurfaceContextHandle()),
+          _device(VK_NULL_HANDLE),
+          _graphicsQueue(VK_NULL_HANDLE),
+          _presentQueue(VK_NULL_HANDLE),
+          _vulkanSwapchain(VK_NULL_HANDLE),
+          _swapChainImages(std::vector<VkImage>{}),
+          _contextIndex(0),
+          _vulkanSwapChainFormat(VkFormat{}),
+          _swapChainExtent(VkExtent2D{}),
+          _renderPass([&]() { return CreateRenderPass(); }),
+          _memoryAllocator([&]() { return CreateMemoryAllocator(); }),
+          _indicesData{},
+          _state()
+    {
+        _logger.logInfoLine("Provided GPU device: " + GetAdapter()->GetName());
+        Initialise();
+        static_cast<void>(_state.Transition(Threading::VolatileState::Initialised));
+        // TODO: This gonna be an issue...?
+        GetSurface()->SizeChanged += [&](auto args) { OnGraphicsSurfaceSizeChanged(args); };
```

Comment:
How do we find out?

Reply:
An aside - the sample application cannot be resized so not sure if this is slated for review later or not.

Reply:
This is a support thing for later once the windowing namespace is fully ported, I don't want graphics suddenly breaking on me.